src/main/java/org/veupathdb/service/eda/ss/model/reducer/StreamIntersectMerger.java

```diff
@@ -1,92 +1,131 @@
 
 
 /**
- * This class is used to take the intersection of two or more streams of sequenced subsets.
+ * This class is used to take the intersection of two or more streams of sequenced longs. The class produces a
+ * stream of output longs, each of which are present in ALL input streams. This behavior is only well defined if
+ * all input streams are sorted.
  */
 public class StreamIntersectMerger implements Iterator<Long> {
-  private final PeekableIterator[] streams;
-  private final int[] nextStream;
-  private Long currentElement;
-  private int currentStreamIdx;
-  private PeekableIterator currentStream;
+  private final PeekableIterator[] peekableIdIndexStreams;
+  private Long nextOutputIndex;
+  private PeekableIterator currentIdIndexStream;
   private boolean hasStarted;
+  private RingLinkedList streamRing;
 
   /**
    * @param sortedStreams Collection of sorted streams to merge by intersection.
    */
   public StreamIntersectMerger(List<Iterator<Long>> sortedStreams) {
-    this.streams = sortedStreams.stream()
+    // If no input streams are provided, it should act as an "empty" Iterator.
+    // If any input stream is "empty", should act as an "empty" Iterator since we are intersecting Iterators.
+    if (sortedStreams.isEmpty() || !sortedStreams.stream().allMatch(Iterator::hasNext)) {
+      hasStarted = true;
+      peekableIdIndexStreams = null;
+      nextOutputIndex = null;
+      return;
+    }
+    // Convert iterators into peekable iterators.
+    this.peekableIdIndexStreams = sortedStreams.stream()
         .map(PeekableIterator::new)
         .toArray(PeekableIterator[]::new);
-    this.currentStream = Arrays.stream(streams)
-        .max(Comparator.comparing(iter -> iter.peek()))
-        .get();
-    this.currentStreamIdx = IntStream.range(0, streams.length)
-        .filter(i -> streams[i] == currentStream)
-        .findFirst()
-        .getAsInt();
-    // Cache a pointer to the nextStream to avoid excessive modulo operations.
-    this.nextStream = IntStream.range(0, streams.length)
-        .map(i -> (i + 1) % streams.length)
-        .toArray();
+    this.streamRing = new RingLinkedList(peekableIdIndexStreams);
+    this.currentIdIndexStream = streamRing.cursor.currentStream;
     hasStarted = false;
   }
 
   @Override
   public boolean hasNext() {
     if (!hasStarted) {
-      setCurrentElement();
+      findNextMatchingIdIndex();
       hasStarted = true;
     }
-    if (currentElement == null) {
-      return false;
-    } else {
-      return true;
-    }
+    return nextOutputIndex != null;
   }
 
   @Override
   public Long next() {
     if (!hasStarted) {
-      setCurrentElement();
+      findNextMatchingIdIndex();
       hasStarted = true;
     }
-    Long next = currentElement;
-    currentStream.next();
-    setCurrentElement();
+    Long next = nextOutputIndex;
+    currentIdIndexStream.next();
+    findNextMatchingIdIndex();
     return next;
   }
 
+  private void findNextMatchingIdIndex() {
```

Comment:
I wrote this up, just to exercise my original thinking about how this would work.  It's more lines of code, but a single continuous loop until all streams match a value.  Not sure if it's simpler or not to understand.  It does pass your existing tests though.  Feel free to use or not.
```
  private void findNextMatchingIdIndex() {
    // set candidate to current stream's value
    Long candidateIdIndex = currentIdIndexStream.peek();
    if (candidateIdIndex == null) {
      nextOutputIndex = null;
      return;
    }
    // value not null; counts as first concurring stream
    int numConcurringStreams = 1;

    // continual loop, trying to match all iterators to the same value
    while (numConcurringStreams < streamRing.size()) {

      // move to the next stream
      currentIdIndexStream = streamRing.advanceCursor().currentStream;

      // find a value >= our candidate
      Long idIndex = currentIdIndexStream.skipUntilMatchesOrExceeds(candidateIdIndex);

      // operations based on retrieved value
      if (idIndex == null) {
        // reached the end of the stream; no more matches ever
        nextOutputIndex = null;
        return;
      }
      else if (idIndex > candidateIdIndex) {
        // current candidate is no good; use the higher value and reset count
        candidateIdIndex = idIndex;
        numConcurringStreams = 1;
      }
      else { // idIndex == candidateIdIndex
        // current stream matches candidate; increment count and continue with this value
        numConcurringStreams++;
      }
    }

    // able to exit loop without returning null first; must have found a match
    nextOutputIndex = candidateIdIndex;
  }
```